sharding/xorValidatorShuffler.go

```diff
@@ -0,0 +1,331 @@
+	rxs := &randXORShuffler{}
+
+	rxs.UpdateParams(nodesShard, nodesMeta, hysteresis, adaptivity)
+
+	return rxs
+}
+
+// UpdateParams updates the shuffler parameters
+// Should be called when new params are agreed through governance
+func (rxs *randXORShuffler) UpdateParams(
+	nodesShard uint32,
+	nodesMeta uint32,
+	hysteresis float32,
+	adaptivity bool,
+) {
+	// TODO: are there constraints we want to enforce? e.g min/max hysteresis
+	shardHysteresis := uint32(float32(nodesShard) * hysteresis)
+	metaHysteresis := uint32(float32(nodesMeta) * hysteresis)
+
+	rxs.mutShufflerParams.Lock()
+	rxs.shardHysteresis = shardHysteresis
+	rxs.metaHysteresis = metaHysteresis
+	rxs.nodesShard = nodesShard
+	rxs.nodesMeta = nodesMeta
+	rxs.adaptivity = adaptivity
+	rxs.mutShufflerParams.Unlock()
+}
+
+// UpdateNodeLists shuffles the nodes and returns the lists with the new nodes configuration
+// The function needs to ensure that:
+//      1.  Old eligible nodes list will have up to shuffleOutThreshold percent nodes shuffled out from each shard
+//      2.  shuffledOutNodes = oldEligibleNodes + waitingListNodes - minNbNodesPerShard (for each shard)
+//      3.  Old waiting nodes list for each shard will be added to the remaining eligible nodes list
+//      4.  The leaving nodes are checked against the new nodes and shuffled out nodes and removed if present from the
+//          pools and leaving nodes list
+//      5.  The remaining new nodes are equally distributed among the existing shards into waiting lists
+//      6.  The remaining shuffled out nodes are distributed among the existing shards into waiting lists.
+//          We may have three situations:
+//          a)  In case (shuffled out nodes + new nodes) > (nbShards * perShardHysteresis + minNodesPerShard) then
+//              we need to prepare for a split event, so a higher percentage of nodes need to be directed to the shard
+//              that will be split.
+//          b)  In case (shuffled out nodes + new nodes) < (nbShards * perShardHysteresis) then we need to prepare for a
+//              shard merge event so
+//          c)  No change in the number of shards then nothing extra needs to be done
+func (rxs *randXORShuffler) UpdateNodeLists(args UpdateNodesArgs) (map[uint32][]Validator, map[uint32][]Validator) {
+	var shuffledOutNodes []Validator
+	var newEligible map[uint32][]Validator
+	var newWaiting map[uint32][]Validator
+
+	newNbShards := rxs.computeNewShards(args.eligible, args.waiting, args.newNodes, args.leaving, args.nbShards)
+
+	rxs.mutShufflerParams.RLock()
+	canSplit := rxs.adaptivity && newNbShards > args.nbShards
+	canMerge := rxs.adaptivity && newNbShards < args.nbShards
+	rxs.mutShufflerParams.RUnlock()
+
+	if canSplit {
+		newEligible, newWaiting = rxs.splitShards(args.eligible, args.waiting, newNbShards)
+	}
+	if canMerge {
+		newEligible, newWaiting = rxs.mergeShards(args.eligible, args.waiting, newNbShards)
+	}
+
+	for shard, vList := range newWaiting {
```

Comment:
Actually the conditions are per shard, but maybe not very obvious. I will try to explain

If we would end up with less than the supported number of nodes for the current number of shards we would do the shard merge first and then proceed to remove the leaving from the resulting number of shards.

Let's say there are some nodes that announced their intent to leave.
Some of these nodes are in shard0, some in shard1, ..., some in metachain

From every shard we can remove the "leaving" nodes only if there are enough nodes left in the waiting list to take their place and not end up with a number of nodes below minimum accepted (nodesPerShard).

So the iteration is on each shard, search for each node in the leaving slice, and remove them if found as long as there is a node in the waiting list assigned to the shard.

Each node is unique so it can only be found in one shard.

There is a bias indeed for the first nodes in the leaving list, not in the first checked shard, but we can have the list shuffled with the same randomness if needed.

Reply:
ok